### PR TITLE
cd: push tags

### PR DESCRIPTION
**Pull request by cristianmtr**
Add a tag when pushing to Jina Hub.

It only pushes that version if the version doesn't exist.

What this means:

- once you have a version you want to release, make a new git tag : `git tag v0.X`, e.g.
- push tag `git push --tags`
- trigger workflow
- this will also push to `latest`

During CI/CD

- you will only push to `latest`
- if you add a new tag, it will push to that tag AND to `latest`

### CODE REVIEW

**Comment by mapleeit** [comment]
What I'm thinking is setting two workflows:
1. Triggered when master push
    1. Generating a tag based on commit message and previous tag (semantic release)
2. Triggered when a tag is pushed
    1. jina hub push
    
Example:
1. https://github.com/jina-ai/hubble/blob/master/.github/workflows/push-master.yml
2. https://github.com/jina-ai/hubble/blob/master/.github/workflows/deploy-on-tag.yml

Do you think it's better?

File: .github/workflows/scripts/push.sh [hunk 1]
```diff
@@ -35,4 +42,24 @@ echo SECRET=`head -c 3 <(echo $exec_secret)`
...
+if [ -z "$GIT_TAG" ]
+then
+  echo WARNING, no git tag!
+else
+  echo git tag = $GIT_TAG
+  jina hub pull jinahub+docker://$exec_name/$GIT_TAG
+  exists=$?
+  if [[ $exists == 1 ]]; then
+    echo does NOT exist, pushing
+    jina hub push --force $exec_uuid --secret $exec_secret . -t $GIT_TAG
```
Comment (mapleeit) [inline .github/workflows/scripts/push.sh:58]:
```suggestion
    jina hub push --force $exec_uuid --secret $exec_secret . -t $GIT_TAG -t latest
```

File: .github/workflows/scripts/push.sh [hunk 2]
```diff
@@ -35,4 +42,24 @@ echo SECRET=`head -c 3 <(echo $exec_secret)`
...
+  else
+    echo exists, will NOT push
```
Comment (mapleeit) [inline .github/workflows/scripts/push.sh:60]:
```suggestion
    echo exists, only push to latest
    jina hub push --force $exec_uuid --secret $exec_secret .
```

File: .github/workflows/scripts/push.sh [hunk 3]
```diff
@@ -35,4 +42,24 @@ echo SECRET=`head -c 3 <(echo $exec_secret)`
...
+  fi
+fi
+
+# we push to latest every time
+jina hub push --force $exec_uuid --secret $exec_secret .
```
Comment (mapleeit) [inline .github/workflows/scripts/push.sh:65]:
```suggestion

```

File: .github/workflows/scripts/push.sh [hunk 4]
```diff
@@ -35,4 +42,24 @@ echo SECRET=`head -c 3 <(echo $exec_secret)`
...
+
+if [ -z "$GIT_TAG" ]
+then
+  echo WARNING, no git tag!
+else
+  echo git tag = $GIT_TAG
+  jina hub pull jinahub+docker://$exec_name/$GIT_TAG
+  exists=$?
+  if [[ $exists == 1 ]]; then
+    echo does NOT exist, pushing
```
Comment (mapleeit) [inline .github/workflows/scripts/push.sh:57]:
```suggestion
    echo does NOT exist, pushing to latest and $GIT_TAG
```

**Review by mapleeit — CHANGES_REQUESTED** [review]
There is a optimization which can speed up the hub pushing.

**Comment by cristianmtr** [comment]
> What I'm thinking is setting two workflows:
> 
> 1. Triggered when master push
>    
>    1. Generating a tag based on commit message and previous tag (semantic release)
> 2. Triggered when a tag is pushed
>    
>    1. jina hub push

So what this offers is another version, in-between versions, on every PR merged. Could be useful, yeah.

**Comment by mapleeit** [comment]
As for this development pattern. I believe it follows something that called [trunck-based development](https://trunkbaseddevelopment.com/). Which basically means the code in master should always in sync with release code.

**Comment by cristianmtr** [comment]
https://github.com/jina-ai/internal-tasks/issues/185

File: .github/workflows/cd.yml [hunk 5]
```diff
@@ -5,19 +5,30 @@ on:
...
     runs-on: ubuntu-latest
-    container:
-      image: jinaai/jina:master-standard
-      options: "--entrypoint /bin/bash"
     steps:
+      - name: install git
+        run: |
+          # checkout needs git >= 2.18 to check out repo as git repo and not archive
+          # repo is needed for getting tag
+          sudo apt-get update && sudo apt-get install -y git && sudo apt-get upgrade git
```
Comment (WyriHaximus) [inline .github/workflows/cd.yml:18]: FYI `ubuntu-latest` should come with Git installed. Unless they changed that

Comment (cristianmtr) [reply .github/workflows/cd.yml]: Seems it does. The problem was that we were using a custom container. Fixed now. Thanks :)

**Review by mapleeit — DISMISSED** [review]
LGTM

**Review by mapleeit — APPROVED** [review]

**Pull request merged**